Add an option to show the active title in the title

leoetlino opened the pull request:
This can be useful for figuring out at a glance what title is running, along with its game ID / title ID.

The first commit moves the title database code out of DolphinWX, so it can be reused in more than just the UI code.

The second commit allows us to print user friendly names for titles out of the box, without requiring the user to manually download a title database and place it in the correct location. No auto-update mechanism is implemented as it would be substantially more complicated and overkill (imo), since titles are not updated that often. This is okay as the database can be freely used as long as we link back to GameTDB.

This option defaults to on. The title is appended to the end of the title.

![Wii Menu](https://i.imgur.com/UYRe6l4.png)
![Mii Channel](https://i.imgur.com/uMIXmTl.png)

JosJuice commented on Source/Core/Common/TitleDatabase.cpp:
```diff
@@ -0,0 +1,72 @@
+  auto iterator = m_title_map.find(game_id);
+
+  // The title database may not have entries for the full game ID (6 characters),
+  // so we should also search for a matching ID with only four characters.
```
This approach makes it impossible to distinguish between e.g. Mario Kart Wii and the Mario Kart Channel. (See PR #2242.) There needs to be a parameter to this function that says whether the title is a disc or a channel.

JosJuice commented on Source/Core/DolphinWX/ISOFile.h:
```diff
@@ -14,6 +14,11 @@
+{
+class TitleDatabase;
+}
+
```
You can also remove the `unordered_map` include.

leoetlino commented:
>Common is supposed to be for things that aren't specific to GameCube/Wii emulation. TitleDatabase would fit better in DiscIO.

Well, I would argue that TitleDatabase is *not* specific to GC/Wii emulation -- I can easily see something like it be useful outside of an emulator, and as far as I know, the whole format was not made for or used by emulators.

I wouldn't mind moving it out of Common though. However, it doesn't seem to really fit in DiscIO, since we're not reading any emulated container format.

>Using a Dolphin-supplied wiitdb.txt out of the box is unacceptable, because the language it uses for game titles might not be the language the user wants.

Oops, I hadn't thought about other languages (using Dolphin in English :P). How would you suggest fixing this? Could we ship with the database in the other languages? They're not that huge, so I think this is viable.

lioncash commented on Source/Core/Common/TitleDatabase.h:
```diff
@@ -0,0 +1,32 @@
+
+  std::unordered_map<std::string, std::string> m_title_map;
+};
+}
```
`}  // namespace Common`

lioncash commented on Source/Core/Core/ConfigManager.cpp:
```diff
@@ -775,19 +778,27 @@ void SConfig::SetRunningGameMetadata(const std::string& game_id, u64 title_id, u
+  if (game_id == "00000000")
   {
-    NOTICE_LOG(BOOT, "Game ID set to %s", game_id.c_str());
+    m_title_description = "";
```
`m_title_description.clear()`

JosJuice commented:
> Well, I would argue that TitleDatabase is not specific to GC/Wii emulation -- I can easily see something like it be useful outside of an emulator, and as far as I know, the whole format was not made for or used by emulators.

TitleDatabase is hardcoding things like the "wiitdb.txt" name, the difference between six-letter and four letter IDs, the system menu ID... That's enough to make it unsuitable for other applications.

> Oops, I hadn't thought about other languages (using Dolphin in English :P). How would you suggest fixing this? Could we ship with the database in the other languages? They're not that huge, so I think this is viable.

If the different language versions of the database all have correct data, it would probably be fine to ship all of them and pick one based on the user's settings. It would just require a bit more storage space.

JosJuice commented on Source/Core/Common/TitleDatabase.cpp:
```diff
@@ -0,0 +1,72 @@
+  // Titles that cannot be part of the Wii TDB,
+  // but common enough to justify having entries for them.
+  m_title_map.emplace("0000000100000002", "System Menu");
+  m_title_map.emplace("OHBC", "The Homebrew Channel");
```
Apparently wiitdb.txt lists HBC as JODI and LULZ but not any other ID. Odd.

lioncash commented on Source/Core/Common/TitleDatabase.cpp:
```diff
@@ -0,0 +1,72 @@
+    return game_id;
+  return StringFromFormat("%s (%s)", title_name.c_str(), game_id.c_str());
+}
+}
```
`}  // namespace Common`

lioncash commented on Source/Core/Common/TitleDatabase.cpp:
```diff
@@ -0,0 +1,72 @@
+
+#include "Common/FileUtil.h"
+#include "Common/StringUtil.h"
+#include "Common/TitleDatabase.h"
```
This should be placed as the first include in the file (e.g. see [Core.cpp](https://github.com/dolphin-emu/dolphin/blob/master/Source/Core/Core/Core.cpp#L5)) as it prevents any includes above it from satisfying any dependencies in the header indirectly if they're missing.

lioncash commented on Source/Core/Common/TitleDatabase.h:
```diff
@@ -0,0 +1,32 @@
+
+namespace Common
+{
+class TitleDatabase final
```
This should probably have a small description indicating what its purpose is.

leoetlino commented:
Moved TitleDatabase to Core/, and fixed everything mentioned by Lioncash.

We now load the appropriate database depending on the interface language, and fall back to English. **But** there is still an issue with this approach: when the user just wants to use the system language, we can't figure out what the current locale code is and load the proper file...

leoetlino commented:
Okay, I have changed it to use the console language.

JosJuice commented on Source/Core/Core/TitleDatabase.cpp:
```diff
@@ -0,0 +1,107 @@
+  m_title_map.emplace("0000000100000002", GetStringT("Wii Menu"));
+  m_title_map.emplace("OHBC", "The Homebrew Channel");
+
+  const std::string language = GetLanguageCode(SConfig::GetInstance().GetCurrentLanguage(true));
```
We should preferably use the GC setting for GC games and Wii setting for Wii games.

leoetlino replied:
Okay, fixed.

JosJuice commented on Source/Core/Core/TitleDatabase.cpp:
```diff
@@ -0,0 +1,107 @@
+  // i18n: "Wii Menu" (or System Menu) refers to the Wii's main menu,
+  // which is (usually) the first thing users see when a Wii console starts.
+  m_title_map.emplace("0000000100000002", GetStringT("Wii Menu"));
+  m_title_map.emplace("OHBC", "The Homebrew Channel");
```
I noticed that you added a GetStringT for the Wii Menu, but doesn't HBC have localized names too?

And my earlier comment about the IDs of HBC hasn't been addressed.

leoetlino replied:
I'm not sure the HBC should be localised, tbh. I don't know about other languages, but in French it's referred to as "the Homebrew Channel" most of the time. Since the artwork (channel banner + in-channel graphics) only says "the Homebrew Channel", I wouldn't be surprised if this was the case in other languages though. If you know this isn't the case, let me know and I'll add the GetStringT.

I'll add all the other IDs of the HBC to the default title map, just in case.

JosJuice replied:
I don't know how most users refer to it, but the channel name is localized in the Wii menu:

![0000000100000002-1](https://cloud.githubusercontent.com/assets/6716818/26121851/e5bb1c40-3a75-11e7-84b6-6b3ec8e0f9dc.png)
![0000000100000002-2](https://cloud.githubusercontent.com/assets/6716818/26121856/e8803a64-3a75-11e7-83bb-090ba69a5a13.png)

But either way, maybe we should avoid adding a GetStringT for this, since it uses the interface language rather than the console language, leading to inconsistency. Since wiitdb.txt already has two IDs for the Homebrew Channel, I don't see why it can't have the other IDs too.

leoetlino replied:
Ah, fair enough. I wasn't aware the channel title *was* localised.

You mean we should avoid having the GetStringT for the Wii Menu too?

JosJuice replied:
> You mean we should avoid having the GetStringT for the Wii Menu too?

I'm not sure. I suppose doing it with GetStringT is better than not localizing it at all, though (and it's less confusing for the Wii Menu to be inconsistent since it's a bit of a "special" title - you can't see its name in the Wii Menu's list of channels.)

leoetlino replied:
Yeah, I agree with this. I'll leave it localised, then.

JosJuice commented on Source/Core/Core/ConfigManager.cpp:
```diff
@@ -743,14 +746,15 @@ void SConfig::LoadSettingsFromSysconf()
 void SConfig::ResetRunningGameMetadata()
 {
-  SetRunningGameMetadata("00000000", 0, 0);
+  SetRunningGameMetadata("00000000", 0, 0, Core::TitleDatabase::TitleType::Other);
```
The commit that changes this comes before the commit that added the enum. (Or is GitHub showing commits in an unintuitive order again?)

leoetlino replied:
Yeah, it is. This is the proper commit order, for the reference:

<details>

f8bc392e7c - (HEAD -> show-active-title, fork/show-active-title) UI: Expose the "show active title" setting (52 minutes ago) <Léo Lam>
bdda2c069a - Add an option to show the active title in the title (52 minutes ago) <Léo Lam>
c7df452f5f - WX: Use TitleDatabase for getting title names (52 minutes ago) <Léo Lam>
9aadcd2d60 - Include the latest WiiTDB (52 minutes ago) <Léo Lam>
f535cab404 - Add a TitleDatabase (52 minutes ago) <Léo Lam>

</details>

Each of these commits should even build individually.

JosJuice commented on Data/Sys/wiitdb-en.txt:
```diff
@@ -0,0 +1,9305 @@
+GGZH52 = Madagascar
+GGZI52 = Madagascar
+GGZJB2 = Madagascar
+GGZP52 = M
```
Isn't this supposed to be at the top?

JosJuice replied:
Er, sorry, this highlighted the wrong line. I meant the `TITLES =` line.

leoetlino replied:
This line is normally at the top, but we need the whole file to be sorted for `comm` to work. I don't think this is an issue though, since it is trivial to find this line with something like grep.

JosJuice replied:
It's only easy to find if you already know that there is a line that starts with `TITLE =`. I think it's useful to let people easily see the date of the file even if they aren't actively looking for it.

leoetlino replied:
Fair enough. The timestamp line is now moved back at the top.

JosJuice commented on Source/Core/Core/TitleDatabase.cpp:
```diff
@@ -0,0 +1,149 @@
+  return true;
+}
+
+static bool IsGCTitle(const std::string& game_id)
```
Please add a comment that this function only is supposed to be used with wiitdb.txt format game IDs, not Dolphin format game IDs. (TurboGrafx-16 titles will be misdetected if a Dolphin format game ID is passed in.)

leoetlino replied:
Done.
